File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.py


class Solution:
    def earliestFinishTime(
        self,
        landStartTime: List[int],
        landDuration: List[int],
        waterStartTime: List[int],
        waterDuration: List[int],
    ) -> int:
        
        def get_min_sequential_finish(start_first, duration_first, start_second, duration_second):
            # Phase 1: Find the earliest completion time for the first activity
            earliest_first_finish = inf
            for i in range(len(start_first)):
                earliest_first_finish = min(earliest_first_finish, start_first[i] + duration_first[i])
            
            # Phase 2: Find the earliest completion time for the second activity
            earliest_second_finish = inf
            for i in range(len(start_second)):
                # Second activity can only start after the first activity is finished
                actual_start_time = max(start_second[i], earliest_first_finish)
                earliest_second_finish = min(earliest_second_finish, actual_start_time + duration_second[i])
                
            return earliest_second_finish

        # Strategy A: Land first, then Water
        land_then_water = get_min_sequential_finish(
            landStartTime, landDuration, waterStartTime, waterDuration
        )
        # Strategy B: Water first, then Land
        water_then_land = get_min_sequential_finish(
            waterStartTime, waterDuration, landStartTime, landDuration
        )
        
        return min(land_then_water, water_then_land)

File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.cpp
class Solution {
private:
    long long getMinSequentialFinish(
        vector<int>& startFirst, vector<int>& durationFirst,
        vector<int>& startSecond, vector<int>& durationSecond
    ) {
        // Phase 1: Find the earliest completion time for the first activity
        long long earliestFirstFinish = LLONG_MAX;
        for (size_t i = 0; i < startFirst.size(); ++i) {
            earliestFirstFinish = min(earliestFirstFinish, (long long)startFirst[i] + durationFirst[i]);
        }

        // Phase 2: Find the earliest completion time for the second activity
        long long earliestSecondFinish = LLONG_MAX;
        for (size_t i = 0; i < startSecond.size(); ++i) {
            long long actualStartTime = max((long long)startSecond[i], earliestFirstFinish);
            earliestSecondFinish = min(earliestSecondFinish, actualStartTime + durationSecond[i]);
        }

        return earliestSecondFinish;
    }

public:
    int earliestFinishTime(
        vector<int>& landStartTime, vector<int>& landDuration,
        vector<int>& waterStartTime, vector<int>& waterDuration
    ) {
        long long landThenWater = getMinSequentialFinish(landStartTime, landDuration, waterStartTime, waterDuration);
        long long waterThenLand = getMinSequentialFinish(waterStartTime, waterDuration, landStartTime, landDuration);
        
        return min(landThenWater, waterThenLand);
    }
};